osu.Server.Spectator/Hubs/MultiplayerQueue.cs

```diff
@@ -155,61 +155,107 @@ public async Task AddItem(MultiplayerPlaylistItem item, MultiplayerRoomUser user
         /// Duplicates <see cref="CurrentItem"/> into the database.
         /// </summary>
         /// <param name="db">The database connection.</param>
-        private async Task duplicateCurrentItem(IDatabaseAccess db)
+        private async Task duplicateCurrentItem(IDatabaseAccess db) => await addItem(db, new MultiplayerPlaylistItem
         {
-            var newItem = new MultiplayerPlaylistItem
-            {
-                OwnerID = CurrentItem.OwnerID,
-                BeatmapID = CurrentItem.BeatmapID,
-                BeatmapChecksum = CurrentItem.BeatmapChecksum,
-                RulesetID = CurrentItem.RulesetID,
-                AllowedMods = CurrentItem.AllowedMods,
-                RequiredMods = CurrentItem.RequiredMods
-            };
-
-            newItem.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, newItem));
-            room.Playlist.Add(newItem);
-
-            await hub.OnPlaylistItemAdded(room, newItem);
+            OwnerID = CurrentItem.OwnerID,
+            BeatmapID = CurrentItem.BeatmapID,
+            BeatmapChecksum = CurrentItem.BeatmapChecksum,
+            RulesetID = CurrentItem.RulesetID,
+            AllowedMods = CurrentItem.AllowedMods,
+            RequiredMods = CurrentItem.RequiredMods
+        });
+
+        private async Task addItem(IDatabaseAccess db, MultiplayerPlaylistItem item)
+        {
+            // Add the item to the list with an initial "infinite" ID, in order to notify updatePlaylistItem()
+            // that the item should be de-prioritised and events should not be propagated for it.
```

Comment:
This explains why the ID is set to `MaxValue` (which i'm not sure I agree with yet) but doesn't explain why the playlist order needs to be updated in this order. I'd need to understand why that is required before making a call on whether this `long.MaxValue` logic is something I can get behind or not.

Reply:
I hoped that this being in "addItem" would explain why the item should be deprioritised. Because it's the last item added?

I could just as well have given it a value of `Current_Max_ID + 1` and it would be the same ordering. I'm not sure how I could explain this any better - it's at the end of the list.

Reply:
Is there a reason the order of these operations couldn't be reversed to get the same behaviour? The `playlist_order` does seem to be null initially, so I think that would still be in a manageable state?

Reply:
I don't understand what you mean. Reverse which operations? The adding to list is required for the ordering to happen.

Reply:
Something along the lines of this:
![20211206 123729 (JetBrains Rider-EAP)](https://user-images.githubusercontent.com/191335/144783271-92a5e5d3-76ad-462b-bca5-b9afd9989530.png)

Reply:
I do not want to propagate `PlaylistItemChanged` events before `PlaylistItemAdded` events.

Reply:
so.. 

```csharp

        private async Task addItem(IDatabaseAccess db, MultiplayerPlaylistItem item)
        {
            item.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, item));

            room.Playlist.Add(item);

            await hub.OnPlaylistItemAdded(room, item);

            await updatePlaylistOrder(db);
        }

```

?

Reply:
And so client side the item shows up in the incorrect position and then pops back into place later on? I suppose it's fine otherwise.

Reply:
I think I'd prefer that to this workaround. Seems like something we can clean up in the future. Prefer that the server logic is as simple as possible for now.